privatevcs/validation/allowlist/list.go

```diff
@@ -0,0 +1,66 @@
+package allowlist
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"regexp"
+
+	"github.com/pkg/errors"
+	"github.com/spacelift-io/spcontext"
+
+	"github.com/spacelift-io/vcs-agent/privatevcs/validation"
+)
+
+// List implements a validation strategy based on an explicit, exhaustive,
+// code-based allowlist (the legacy approach).
+type List struct {
+	projectRegexp *regexp.Regexp
+}
+
+// New creates a new allowlist strategy from a project regexp.
+func New(projectRegexp string) (*List, error) {
+	r, err := regexp.Compile(projectRegexp)
+	if err != nil {
+		return nil, errors.Wrapf(err, "couldn't compile project regexp %q", projectRegexp)
+	}
+
+	return &List{projectRegexp: r}, nil
+}
+
+// Validate validates the request and returns an error if the request should be
+// blocked.
+func (l List) Validate(ctx *spcontext.Context, vendor validation.Vendor, req *http.Request) (*spcontext.Context, error) {
+	name, project, subdomain, err := MatchRequest(vendor, req)
+	if err != nil {
+		return ctx.With("match_error", err), ctx.RawError(err, "invalid request")
+	}
+
+	if subdomain != nil {
```

Comment:
This bit is a complete pain unfortunately. What happens here is that the VCS agent has the domain name for the VCS hard-coded, so we always just apply the path to the hostname. But if you have GitHub Enterprise subdomain isolation enabled, we actually need to make a request to `codeload.<my-github-domain>` rather than just `<my-github-domain>`. Here's the commit where I added this change: https://github.com/spacelift-io/vcs-agent/commit/ccf62feb2f2bfc1f3954261ba09b6be46f0ca623

TL;DR; I think we might need to pull this code out of the validation rules and move it into agent.go, and just always do this check before making a request (i.e. if the vendor is GitHub and we're trying to download the tarball, modify the hostname if necessary). Otherwise source downloads will break for GitHub Enterprise when not using the allowlist.

Reply:
Done.